src/backend/columnar/columnar_reader.c

```diff
@@ -84,10 +84,25 @@ struct ColumnarReadState
 	int64 chunkGroupsFiltered;
 };
 
+struct ColumnarRandomReadState
+{
+	/*
+	 * StripeReadState cache for random read operations. Note that we don't
+	 * cache the last chunk group here since cachedStripeReadState already
+	 * stores chunkGroupReadState and chunkGroupIndex.
+	 */
+	StripeReadState *cachedStripeReadState;
+
+	/* StripeMetadata of the stripe which we cached the read state of */
+	StripeMetadata *cachedStripeMetadata;
+
+	/* MemoryContext that needs to be used during random read operations */
+	MemoryContext stripeReadContext;
```

Comment:
This feels redundant. Isn't stripeReadContext the same as cachedStripeReadState->stripeReadContext? And isn't cachedStripeMetadata the same as the cachedStripeReadState->currentStripe'th element of cachedStripeReadState->stripeList?

I was imagining that this change would be mostly just a way to skip ahead in a ColumnarReadState if necessary. Is there a reason that won't work?

Reply:
> This feels redundant. Isn't stripeReadContext the same as cachedStripeReadState->stripeReadContext? And isn't cachedStripeMetadata the same as the cachedStripeReadState->currentStripe'th element of cachedStripeReadState->stripeList?

Indeed both is true.

> I was imagining that this change would be mostly just a way to skip ahead in a ColumnarReadState if necessary. Is there a reason that won't work?

I think this should work, let me work little more on those changes.